es-app/src/FileData.cpp

```diff
@@ -185,3 +192,67 @@ void FileData::sort(const SortType& type)
 {
 	sort(*type.comparisonFunction, type.ascending);
 }
+
+void FileData::launchGame(Window* window)
+{
+	LOG(LogInfo) << "Attempting to launch game...";
+
+	AudioManager::getInstance()->deinit();
+	VolumeControl::getInstance()->deinit();
+	window->deinit();
+
+	std::string command = mEnvData->mLaunchCommand;
+
+	const std::string rom = escapePath(getPath());
+	const std::string basename = getPath().stem().string();
+	const std::string rom_raw = fs::path(getPath()).make_preferred().string();
+
+	command = strreplace(command, "%ROM%", rom);
+	command = strreplace(command, "%BASENAME%", basename);
+	command = strreplace(command, "%ROM_RAW%", rom_raw);
+
+	LOG(LogInfo) << "	" << command;
+	int exitCode = runSystemCommand(command);
+
+	if(exitCode != 0)
+	{
+		LOG(LogWarning) << "...launch terminated with nonzero exit code " << exitCode << "!";
+	}
+
+	window->init();
+	VolumeControl::getInstance()->init();
+	AudioManager::getInstance()->init();
+	window->normalizeNextUpdate();
+
+	//update number of times the game has been launched
+
+	FileData* gameToUpdate;
+	if (mIsVirtual)
+	{
+		gameToUpdate = mSourceFileData;
+	}
+	else
+	{
+		gameToUpdate = this;
+	}
+
+	int timesPlayed = gameToUpdate->metadata.getInt("playcount") + 1;
+	gameToUpdate->metadata.set("playcount", std::to_string(static_cast<long long>(timesPlayed)));
+
+	//update last played time
+	boost::posix_time::ptime time = boost::posix_time::second_clock::universal_time();
+	gameToUpdate->metadata.setTime("lastplayed", time);
+	LOG(LogDebug) << "Finished updating data. Going to update virtual Files. ";
+	VirtualSystemManager::get()->updateVirtualSystems(gameToUpdate);
+}
+
+void FileData::refreshMetadata()
+{
+	if(mIsVirtual)
+	{
+		metadata = mSourceFileData->metadata;
+		mVirtualName = removeParenthesis(mSourceFileData->metadata.get("name"));
+		boost::trim(mVirtualName);
+		mVirtualName += " [" + mSourceFileData->getSystem()->getName() + "]";
+	}
```

Comment:
The virtual name is used as the display name for Virtual entries. We are _appending_ the system name as this is what the user sees in the list, so that in a multi-system collection the user can identify what games are from which system, and so that the user can play the version of the game they want - think of simple clashes like Alladin SNES/MegaDrive, but even games that showed up in more systems than I can count. 

I imagine one could also have a theme element per entry to mark it as belonging to a specific system, but that's beyond the scope (and design discussion) of this current change - similar to how to represent Favorites in the UI.

See:

https://youtu.be/rUvua2yPIyQ?t=10